torch/csrc/autograd/variable.cpp

```diff
@@ -106,52 +105,71 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, int output_
   attr_version = version_counter.current_version();
 }
 
-std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
+std::shared_ptr<Function>& Variable::ViewImpl::get_grad_fn() {
   std::lock_guard<std::mutex> lock(mutex);
-  if (!_grad_fn && !base.requires_grad()) {
-    return _grad_fn;
+  if (!grad_fn && !base.requires_grad()) {
+    return grad_fn;
   }
   auto current_version = version_counter.current_version();
   if (attr_version != current_version) {
     TORCH_ASSERT(output_nr == 0);
     auto fn = std::make_shared<generated::AsStridedBackward>();
-    fn->self_geometry = TensorGeometry(base);
+    fn->self_geometry = at::TensorGeometry(base);
     fn->size = sizes();
     fn->stride = strides();
     fn->storage_offset = data.storage_offset();
     fn->set_next_functions(get_next_functions(base));
     fn->num_inputs = 1;
-    _grad_fn = std::move(fn);
+    grad_fn = std::move(fn);
     attr_version = current_version;
   }
-  return _grad_fn;
+  return grad_fn;
 }
 
-void VariableViewImpl::rebase_history(int output_nr, std::shared_ptr<Function> grad_fn) {
-  TORCH_ASSERT(output_nr == 0);
-  TORCH_ASSERT(grad_fn);
-  TORCH_ASSERTM(grad_fn->num_inputs == 1, "Functions which modify views in-place must return a single Variable");
-  this->output_nr = output_nr;
-  base.output_nr() = 0;
-  base.get()->_grad_fn = std::make_shared<CopySlices>(
-      base, TensorGeometry(data), std::move(grad_fn));
-  get_grad_fn();  // trigger an update to the view's grad_fn
+void Variable::ViewImpl::rebase_history(Edge gradient_edge) {
+  TORCH_ASSERT(gradient_edge.input_nr == 0);
+  TORCH_ASSERT(gradient_edge.function);
+  TORCH_ASSERTM(
+      gradient_edge.function->num_inputs == 1,
+      "Functions which modify views in-place must return a single Variable");
+  this->output_nr = gradient_edge.input_nr;
+  auto copy_slices = std::make_shared<CopySlices>(
+      base, at::TensorGeometry(data), std::move(gradient_edge.function));
+  base.set_gradient_edge({std::move(copy_slices), 0});
+  get_grad_fn(); // trigger an update to the view's grad_fn
+}
+
+void Variable::rebase_history(Edge gradient_edge) {
+  TORCH_ASSERT(gradient_edge.function != nullptr);
+  if (is_view()) {
+    auto& impl = static_cast<Variable::ViewImpl&>(*get());
+    impl.rebase_history(std::move(gradient_edge));
+  } else {
+    set_gradient_edge(std::move(gradient_edge));
+  }
 }
 
 Variable Variable::detach() const {
-  Variable detached = make_variable(data());
-  detached.version_counter() = version_counter();
+  auto detached = make_variable(data(), /*requires_grad=*/false);
+  detached.set_version(version_counter());
```

Comment:
(irrelevant to this PR) @colesbury do you remember why do we detach the version counter as well? I don't think this should ever happen, since that will break correctness checks

Reply:
We share the version counter when we create a detached Variable since they share the same data. That way, if you modify the detached variable, you'll still get the error message in backwards if it messes up a SavedVariable.

Is that what you mean by "why do we detach the version counter as well"?

Reply:
Ugh I read `version_counter()` as a call to the constructor, not `this->version_counter()`. Sorry...